server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/model/Capability.scala

```diff
@@ -0,0 +1,93 @@
+}
+
+sealed trait CapabilityProperties
+
+sealed trait Capability {
+  def identifier(): CapabilityIdentifier
+  def properties(): CapabilityProperties
+}
+
+final case class CoreCapability(identifier: CapabilityIdentifier = JMAP_CORE,
+                                properties: CoreCapabilityProperties) extends Capability
+
+object CoreCapabilityProperties {
+  def apply(maxSizeUpload: UnsignedInt, maxConcurrentUpload: UnsignedInt,
+            maxSizeRequest: UnsignedInt, maxConcurrentRequests: UnsignedInt,
+            maxCallsInRequest: UnsignedInt, maxObjectsInGet: UnsignedInt,
+            maxObjectsInSet: UnsignedInt, collationAlgorithms: List[String]): CoreCapabilityProperties = {
+
+    require(Option(maxSizeUpload).isDefined, "maxSizeUpload cannot be null")
+    require(Option(maxConcurrentUpload).isDefined, "maxConcurrentUpload cannot be null")
+    require(Option(maxSizeRequest).isDefined, "maxSizeRequest cannot be null")
+    require(Option(maxConcurrentRequests).isDefined, "maxConcurrentRequests cannot be null")
+    require(Option(maxCallsInRequest).isDefined, "maxCallsInRequest cannot be null")
+    require(Option(maxObjectsInGet).isDefined, "maxObjectsInGet cannot be null")
+    require(Option(maxObjectsInSet).isDefined, "maxObjectsInSet cannot be null")
+    require(Option(collationAlgorithms).isDefined, "collationAlgorithms cannot be null")
+
+    new CoreCapabilityProperties(maxSizeUpload, maxConcurrentUpload, maxSizeRequest, maxConcurrentRequests, maxCallsInRequest,
+      maxObjectsInGet, maxObjectsInSet, collationAlgorithms)
+  }
+}
+
+final case class CoreCapabilityProperties private(maxSizeUpload: UnsignedInt, maxConcurrentUpload: UnsignedInt,
+                                                  maxSizeRequest: UnsignedInt, maxConcurrentRequests: UnsignedInt,
+                                                  maxCallsInRequest: UnsignedInt, maxObjectsInGet: UnsignedInt,
+                                                  maxObjectsInSet: UnsignedInt, collationAlgorithms: List[String]) extends CapabilityProperties {
+}
+
+final case class MailCapability(identifier: CapabilityIdentifier = JMAP_MAIL,
+                                properties: MailCapabilityProperties) extends Capability
+
+object MailCapabilityProperties {
+  def apply(maxMailboxesPerEmail: Option[UnsignedInt], maxMailboxDepth: Option[UnsignedInt],
+            maxSizeMailboxName: UnsignedInt, maxSizeAttachmentsPerEmail: UnsignedInt,
+            emailQuerySortOptions: List[String], mayCreateTopLevelMailbox: Boolean): MailCapabilityProperties = {
+    require(Option(maxSizeMailboxName).isDefined, "maxSizeMailboxName cannot be null")
+    require(Option(maxSizeAttachmentsPerEmail).isDefined, "maxSizeAttachmentsPerEmail cannot be null")
+    require(Option(emailQuerySortOptions).isDefined, "emailQuerySortOptions cannot be null")
+
+    new MailCapabilityProperties(maxMailboxesPerEmail, maxMailboxDepth, maxSizeMailboxName, maxSizeAttachmentsPerEmail,
```

Comment:
I bet you don't need new

Reply:
`new MailCapabilityProperties()` is calling case class constructor. `MailCapabilityProperties()` is calling object MailCapabilityProperties `apply()` method. If I remove `new`, there will be endless recursive call.

Because now I split case class into companion classes. Idem for other same remarks